pkg/cmd/server/kubernetes/master_config.go

```diff
@@ -205,39 +217,48 @@ func BuildKubernetesMasterConfig(options configapi.MasterConfig, requestContextM
-	// storageVersions is a map from API group to allowed versions that must be a version exposed by the REST API or it breaks.
-	// We need to fix the upstream to stop using the storage version as a preferred api version.
-	storageVersions := map[string]string{}
-
-	enabledKubeVersions := configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupKube)
-	if len(enabledKubeVersions) > 0 {
-		kubeStorageVersion := unversioned.GroupVersion{Group: configapi.APIGroupKube, Version: options.EtcdStorageConfig.KubernetesStorageVersion}
-		databaseStorage, err := NewEtcdStorage(etcdClient, kubeStorageVersion, options.EtcdStorageConfig.KubernetesStoragePrefix)
-		if err != nil {
-			return nil, fmt.Errorf("Error setting up Kubernetes server storage: %v", err)
-		}
-		storageDestinations.AddAPIGroup(configapi.APIGroupKube, databaseStorage)
-		storageVersions[configapi.APIGroupKube] = options.EtcdStorageConfig.KubernetesStorageVersion
+	resourceEncodingConfig := genericapiserver.NewDefaultResourceEncodingConfig()
+	resourceEncodingConfig.SetVersionEncoding(
+		configapi.APIGroupKube,
+		unversioned.GroupVersion{Group: configapi.APIGroupKube, Version: options.EtcdStorageConfig.KubernetesStorageVersion},
+		configapi.SchemeGroupVersion,
+	)
+
+	if len(configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupExtensions)) > 0 {
+		resourceEncodingConfig.SetVersionEncoding(
+			configapi.APIGroupExtensions,
+			unversioned.GroupVersion{Group: extensions.GroupName, Version: "v1beta1"},
+			extensions.SchemeGroupVersion,
+		)
 	}
 
-	// enable this if extensions API is enabled (or batch or autoscaling, since they persist to extensions/v1beta1 for now)
-	// TODO: replace this with a loop over configured storage versions
-	extensionsEnabled := len(configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupExtensions)) > 0
-	batchEnabled := len(configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupBatch)) > 0
-	autoscalingEnabled := len(configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupAutoscaling)) > 0
-	if extensionsEnabled || autoscalingEnabled || batchEnabled {
-		// TODO: replace this with a configured storage version for extensions once configuration exposes this
-		extensionsStorageVersion := unversioned.GroupVersion{Group: extensions.GroupName, Version: "v1beta1"}
-		databaseStorage, err := NewEtcdStorage(etcdClient, extensionsStorageVersion, options.EtcdStorageConfig.KubernetesStoragePrefix)
-		if err != nil {
-			return nil, fmt.Errorf("Error setting up Kubernetes extensions server storage: %v", err)
-		}
-		storageDestinations.AddAPIGroup(configapi.APIGroupExtensions, databaseStorage)
-		storageVersions[configapi.APIGroupExtensions] = extensionsStorageVersion.String()
+	if len(configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupBatch)) > 0 {
```

Comment:
I think I'd prefer the storage version configuration not be conditional on the REST API's being enabled... we should know how we want to store/retrieve those even if they're not surfaced in the API